File: monsoon3.0/src/main/java/wtf/monsoon/api/setting/Setting.java
package wtf.monsoon.api.setting;

import lombok.Getter;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.Supplier;

import wtf.monsoon.Wrapper;
import wtf.monsoon.impl.event.EventUpdateEnumSetting;

public class Setting<T> {

    @Getter private final String name;
    @Getter private String description = "";

    private Supplier<Boolean> visibility = () -> true;

    @Getter private Setting<?> parent = null;
    @Getter private List<Setting<?>> children = new ArrayList<>();

    @Getter protected T value;
    @Getter private T defaultValue;

    @Getter private T minimum;
    @Getter private T maximum;
    @Getter private T incrementation;

    @Getter private int index = 0;

    public Setting(String name, T value) {
        this.name = name;
        this.value = value;
        this.defaultValue = value;

        if (getValue() instanceof Enum<?>) {
            index = ((Enum<?>) value).ordinal();
        }
    }

    public void setValue(T value) {
        if (value instanceof Enum<?>) {
            EventUpdateEnumSetting event = new EventUpdateEnumSetting((Setting<Enum<?>>) this, (Enum<?>) this.value, (Enum<?>) value);

            Wrapper.getEventBus().post(event);

            if (!event.isCancelled()) {
                this.value = value;
            }
        } else {
            this.value = value;
        }
    }

    public void setValueSilent(T value) {
        this.value = value;
    }

    public T getMode(boolean previous) {
        if (getValue() instanceof Enum<?>) {
            Enum<?> enumeration = (Enum<?>) getValue();

            String[] values = Arrays.stream(enumeration.getClass().getEnumConstants()).map(Enum::name).toArray(String[]::new);

            if (!previous) {
                index = index + 1 > values.length - 1 ? 0 : index + 1;
            } else {
                index = index - 1 < 0 ? values.length - 1 : index - 1;
            }

            return (T) Enum.valueOf(enumeration.getClass(), values[index]);
        }

        return null;
    }

    public Setting<T> minimum(T minimum) {
        this.minimum = minimum;
        return this;
    }

    public Setting<T> maximum(T maximum) {
        this.maximum = maximum;
        return this;
    }

    public Setting<T> incrementation(T incrementation) {
        this.incrementation = incrementation;
        return this;
    }

    public Setting<T> describedBy(String description) {
        this.description = description;
        return this;
    }

    public Setting<T> visibleWhen(Supplier<Boolean> visibility) {
        this.visibility = visibility;
        return this;
    }

    public Setting<T> childOf(Setting<?> parent) {
        this.parent = parent;
        this.parent.children.add(this);

        return this;
    }

    public String getPath() {
        return getParent() == null ? getName() : getParent().getPath() + getName();
    }

    public List<Setting<?>> getHierarchy() {
        List<Setting<?>> hierarchy = new ArrayList<>();

        for (Setting<?> subsetting : getChildren()) {
            hierarchy.add(subsetting);
            hierarchy.addAll(subsetting.getHierarchy());
        }

        return hierarchy;
    }

    public boolean isVisible() {
        return visibility.get();
    }

}

File: monsoon-3.x-v2/src/main/java/wtf/monsoon/backend/setting/Setting.kt
package wtf.monsoon.backend.setting

import spritz.api.annotations.Excluded
import spritz.api.annotations.Identifier
import wtf.monsoon.backend.Feature
import wtf.monsoon.backend.module.mode.ModuleMode

/**
 * @author surge
 * @since 09/02/2023
 */
class Setting<T>(name: String, description: String, private var value: T) : Feature(name, description) {

    @Excluded val default = value

    @Excluded var minimum: T? = null
    @Excluded var maximum: T? = null
    @Excluded var incrementation: T? = null

    private var index: Int = 0

    @Excluded var parent: Setting<*>? = null
    @Excluded val children = mutableListOf<Setting<*>>()

    @Excluded var modes: List<ModuleMode<*>>? = null

    @Excluded var visibility: () -> Boolean = { true }

    constructor(name: String, description: String, value: T, minimum: T, maximum: T, incrementation: T) : this(name, description, value) {
        this.minimum = minimum
        this.maximum = maximum
        this.incrementation = incrementation
    }

    constructor(name: String, description: String, modes: List<ModuleMode<*>>) : this(name, description, modes[0] as T) {
        this.modes = modes;
    }

    constructor(name: String, value: T) : this(name, "No description set", value)

    @Identifier("get_value")
    fun getValue(): T {
        return value
    }

    @Identifier("set_value")
    fun setValue(value: T) {
        this.value = value
    }

    @Excluded
    infix fun childOf(parent: Setting<*>): Setting<T> {
        this.parent = parent
        this.parent!!.children.add(this)

        return this
    }

    @Excluded
    infix fun visibleWhen(condition: () -> Boolean): Setting<T> {
        this.visibility = condition
        return this
    }

    @Excluded
    infix fun minimum(minimum: T): Setting<T> {
        this.minimum = minimum
        return this
    }

    @Excluded
    infix fun maximum(maximum: T): Setting<T> {
        this.maximum = maximum
        return this
    }

    @Excluded
    infix fun incrementation(incrementation: T): Setting<T> {
        this.incrementation = incrementation
        return this
    }

    @Excluded
    val nextMode: T
        get() {
            return if(value is Enum<*>) {
                val enum = value as Enum<*>

                java.lang.Enum.valueOf(
                    enum::class.java,
                    enum.javaClass.enumConstants.map { it.name }[nextIndex]
                ) as T
            } else {
                this.modes?.get(nextIndex) as T
            }
        }

    @Excluded
    private val nextIndex: Int
        get() {
            if(value is Enum<*>) {
                val enum = value as Enum<*>

                if (index + 1 > enum.javaClass.enumConstants.size - 1) {
                    index = 0
                } else {
                    index += 1
                }

                return index
            } else {
                if (index + 1 > modes!!.size - 1) {
                    index = 0
                } else {
                    index += 1
                }

                return index
            }
        }

}